bindings/wasm/bindings.js

```diff
@@ -162,6 +170,49 @@ Module.setup = function () {
     return result;
   };
 
+  Module.Manifold.prototype.getMeshDirect = function(opt = { mode: 0 }) {
+    const mode = opt.mode || 0;
+    const vertexFactory = opt.vertexTarget || float32factory;
+    const indexFactory = opt.indexTarget || uint32factory;
+    const mempos = this._GetMeshDirect(mode) / 4;
+    const umem = Module.HEAPU32;
+    const fmem = Module.HEAPF32;
+    const head = umem.subarray(mempos, mempos + 6);
+    const vS = head[0]; // vert start
+    const vE = head[1]; // vert end
+    const vR = head[2]; // num vertex records
+    const eS = head[3]; // index start
+    const eE = head[4]; // index end
+    const eR = head[5]; // num index records
+    const vert = fmem.subarray(vS / 4, vE / 4); // vertex data (direct)
+    const indx = umem.subarray(eS / 4, eE / 4); // index data (direct / interleaved)
+    const vertex = vertexFactory(mode === 2 ? eR * 9 : vR * 3);
+    const index = mode != 2 ? indexFactory(eR * 3) : undefined;
+    switch (mode) {
+      case 0:
+        vertex.set(vert);
+        // 30x slower than for let
+        // index.set(indx.filter( (v,i) => i % 4 === 0));
+        for (let p = 0, i = 0, l = indx.length; i < l; i += 4) {
+            index[p++] = indx[i];
+        }
+        break;
+      case 1:
+        // letting WASM compact the index array is another speed boost
+        vertex.set(vert);
+        index.set(umem.subarray(mempos + 6, mempos + 6 + eR * 3));
+        break;
+      case 2:
+        // in some cases, we want the unrolled vertex array (non-indexed)
+        // doing this work in WASM is substantially faster
+        vertex.set(fmem.subarray(mempos + 6, mempos + 6 + eR * 9));
+        break;
+    }
+    this._FreeMeshDirect(mempos * 4);
+
+    return { vertex, index };
```

Comment:
I would guess that this part: that the JS function returns a flat array, is what is giving the majority of the speedup. Could we try bench-marking just that without the rest of the C++ changes?

Reply:
that's mode 0 (just the heap pointers) vs mode 1 (heap pointer to flat vertex array and malloc'd new flat index array). you can't get either of these without the C++ helper function to return the pointers to the heap locations.

I, too, would prefer a more elegant-looking solution.